### PR TITLE
Bugfix - a task will run multiple times when chaining chains with groups

### LINKED ISSUE #9020: A task will run multiple times when chaining chains with groups
# Checklist

- [x] I have verified that the issue exists against the `main` branch of Celery.
- [ ] This has already been asked to the [discussions forum](https://github.com/celery/celery/discussions) first.
- [x] I have read the relevant section in the
  [contribution guide](https://docs.celeryq.dev/en/main/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/main)
  to find out if the bug was already fixed in the main branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).
- [x] I have tried to reproduce the issue with [pytest-celery](https://docs.celeryq.dev/projects/pytest-celery/en/latest/userguide/celery-bug-report.html) and added the reproduction script below.

## Mandatory Debugging Information

- [ ] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [ ] I have verified that the issue exists against the `main` branch of Celery.
- [ ] I have included the contents of ``pip freeze`` in the issue.
- [ ] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- None

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**:

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

- **Minimal Python Version**: N/A or Unknown
- **Minimal Celery Version**: N/A or Unknown
- **Minimal Kombu Version**: N/A or Unknown
- **Minimal Broker Version**: N/A or Unknown
- **Minimal Result Backend Version**: N/A or Unknown
- **Minimal OS and/or Kernel Version**: N/A or Unknown
- **Minimal Broker Client Version**: N/A or Unknown
- **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
```

</p>
</details>

# Expected Behavior

# Actual Behavior

Happens on the latest version of celery.

On some complex scenarios, when there is some chaining of chains that contain groups, a task that is chained after a group will run multiple times.
For example, here task_b and task_c should be printed only once, but each one is printed twice
```
app = Celery('tasks', broker='memory://', result_backend='cache+memory://')


@app.task(bind=True)
def task_a(self):
    print("task_a")


@app.task(bind=True)
def task_b(self):
    print('task_b')


@app.task(bind=True)
def task_c(self):
    print("task_c")


group1 = celery.group(task_a.si(), task_a.si())
group2 = celery.group(task_a.si(), task_a.si())
chord1 = group1 | group2
chord2 = celery.chain(chord1, (task_b.si() | task_c.si())).apply_async()
```

Output:
```
[2024-05-13 10:12:37,405: INFO/MainProcess] Task celery.task_a[d947d22f-900a-454e-a3cd-dee12dedfb4a] succeeded in 0.5s: None
[2024-05-13 10:12:37,405: INFO/MainProcess] Task celery.task_a[4dc6c550-70fb-46d6-8967-1c92783682b4] received
[2024-05-13 10:12:37,405: WARNING/MainProcess] task_a
[2024-05-13 10:12:37,407: INFO/MainProcess] Task celery.task_a[4dc6c550-70fb-46d6-8967-1c92783682b4] succeeded in 0.0s: None
[2024-05-13 10:12:37,407: INFO/MainProcess] Task celery.task_a[eb8dc8d9-30e6-4b30-aa4f-b867d85b817c] received
[2024-05-13 10:12:37,408: WARNING/MainProcess] task_a
[2024-05-13 10:12:37,408: INFO/MainProcess] Task celery.task_a[eb8dc8d9-30e6-4b30-aa4f-b867d85b817c] succeeded in 0.0s: None
[2024-05-13 10:12:37,409: INFO/MainProcess] Task celery.task_b[88f28e3b-a986-4f20-9ce0-b4a28b0c669b] received
[2024-05-13 10:12:37,409: WARNING/MainProcess] task_b
[2024-05-13 10:12:37,410: INFO/MainProcess] Task celery.task_b[88f28e3b-a986-4f20-9ce0-b4a28b0c669b] succeeded in 0.0s: None
[2024-05-13 10:12:37,410: INFO/MainProcess] Task celery.task_b[88f28e3b-a986-4f20-9ce0-b4a28b0c669b] received
[2024-05-13 10:12:37,410: WARNING/MainProcess] task_b
[2024-05-13 10:12:37,411: INFO/MainProcess] Task celery.task_b[88f28e3b-a986-4f20-9ce0-b4a28b0c669b] succeeded in 0.0s: None
[2024-05-13 10:12:37,411: INFO/MainProcess] Task celery.task_c[8bb0b7fc-c3f8-4fea-a2c9-13e8f9ae7dbd] received
[2024-05-13 10:12:37,411: WARNING/MainProcess] task_c
[2024-05-13 10:12:37,412: INFO/MainProcess] Task celery.task_c[8bb0b7fc-c3f8-4fea-a2c9-13e8f9ae7dbd] succeeded in 0.0s: None
[2024-05-13 10:12:37,412: INFO/MainProcess] Task celery.task_c[8bb0b7fc-c3f8-4fea-a2c9-13e8f9ae7dbd] received
[2024-05-13 10:12:37,412: WARNING/MainProcess] task_c
[2024-05-13 10:12:37,412: INFO/MainProcess] Task celery.task_c[8bb0b7fc-c3f8-4fea-a2c9-13e8f9ae7dbd] succeeded in 0.0s: None
```

### PR DESCRIPTION
**Pull request by DorSSS**
Fixes #9020 

On `_chain.__or__` there is already a logic that knows to take a group/chord, and check if it needs an upgrade to chord. The problem is that if we chain 2 chains, it skips this logic and just constructs a chain from all the tasks, without any checks for an upgrade. Therefore, in a case where the first chain ends with a group, it will not create a chord.

### CODE REVIEW

**Review by auvipy — COMMENTED** [review]
the approach seems reasonable, but I will wait for thorough review from Nusnus

File: celery/canvas.py
Comment (pencil) [inline celery/canvas.py:977]:
```suggestion
```

Seems like this comment is no longer relevant.

**Review by pencil — COMMENTED** [review]
I can confirm that this indeed fixes the test I proposed in #9013 🎉

**Review by auvipy — APPROVED** [review]
approving as both Unit and integrations tests are passing

**Comment by Nusnus** [comment]
> Hey there, sorry for the delay. The fix does look alright, but I’d want to have a smoke test to confirm it 100%.
> 
> Can you please add a test case here? https://github.com/celery/celery/blob/main/t/smoke/tests/test_canvas.py
> 
> LMK if you need any assistance, though it should be straightforward to add a new test (see current tests for example)
> 
> Thank you!

You can run it locally like this:
`tox -e 3.12-smoke -- -k "test_canvas.py and test_chord and YOUR_TEST_FUNC”`

For example:
`tox -e 3.12-smoke -- -k "test_canvas.py and test_chord and test_sanity”`
```console
tox -e 3.12-smoke -- -k "test_canvas.py and test_chord and test_sanity"
ROOT: will run in automatically provisioned tox, host /opt/homebrew/Cellar/tox/4.15.1/libexec/bin/python is missing [requires (has)]: tox-gh-actions
ROOT: provision> /Users/nusnus/dev/GitHub/celery/.tox/.tox/bin/python -m tox -e 3.12-smoke -- -k 'test_canvas.py and test_chord and test_sanity'
ROOT: tox-gh-actions won't override envlist because tox is not running in GitHub Actions
.pkg: _optional_hooks examples/app> python /Users/nusnus/dev/GitHub/celery/.tox/.tox/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
.pkg: get_requires_for_build_editable examples/app> python /Users/nusnus/dev/GitHub/celery/.tox/.tox/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
.pkg: build_editable examples/app> python /Users/nusnus/dev/GitHub/celery/.tox/.tox/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
3.12-smoke: install_package examples/app> python -I -m pip install --force-reinstall --no-deps /Users/nusnus/dev/GitHub/celery/.tox/.tmp/package/48/celery-5.4.0-0.editable-py3-none-any.whl
3.12-smoke: commands[0] examples/app> pytest -xsvv t/smoke --dist=loadscope --reruns 10 --reruns-delay 60 --rerun-except AssertionError -k 'test_canvas.py and test_chord and test_sanity'
======================================================================================= test session starts =======================================================================================
platform darwin -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0 -- /Users/nusnus/dev/GitHub/celery/.tox/3.12-smoke/bin/python
cachedir: .tox/3.12-smoke/.pytest_cache
rootdir: /Users/nusnus/dev/GitHub/celery
configfile: pyproject.toml
plugins: docker-tools-3.1.3, celery-1.0.0, cov-5.0.0, github-actions-annotate-failures-0.2.0, click-1.1.0, subtests-0.12.1, rerunfailures-14.0, order-1.2.1, anyio-4.4.0, timeout-2.3.1, xdist-3.6.1
collected 141 items / 139 deselected / 2 selected

t/smoke/tests/test_canvas.py::test_chord::test_sanity[celery_setup_worker-celery_redis_broker-celery_redis_backend] Creating network pytest-7ef1ad21-ca14-45c7-be01-61b7aaed2089
Fetching redis:latest
Waiting for container to be ready.RedisContainer::interesting_easley is ready.

Creating network pytest-97988f0f-7e32-4dac-88ba-d84c305f6bf6
Building ..........................................................................................
Waiting for container to be ready.RedisContainer::hardcore_nightingale is ready.

Waiting for container to be ready.RedisContainer::dreamy_montalcini is ready.

Creating volume pytest-787ef7ca-ed48-45b4-907a-90e8a2d42d2e
Waiting for container to be ready.Waiting for SmokeWorkerContainer::admiring_hopper to get ready...
SmokeWorkerContainer::admiring_hopper is ready.

PASSED
t/smoke/tests/test_canvas.py::test_chord::test_sanity[celery_setup_worker-celery_rabbitmq_broker-celery_redis_backend] Creating network pytest-8094d5ee-e98e-4d34-9e80-b9c72464d443
Waiting for container to be ready......RabbitMQContainer::dreamy_ritchie is ready.

Waiting for container to be ready.RedisContainer::competent_noyce is ready.

Creating volume pytest-c0e8cf04-f6a5-4259-91f9-c8979c55dc4c
Waiting for container to be ready.Waiting for SmokeWorkerContainer::pensive_mirzakhani to get ready....
SmokeWorkerContainer::pensive_mirzakhani is ready.

PASSED

=============================================================================== 2 passed, 139 deselected in 43.20s ================================================================================
  3.12-smoke: OK (45.02=setup[1.28]+cmd[43.74] seconds)
  congratulations :) (46.79 seconds)
```

**Comment by DorSSS** [comment]
Hi, I tried to create the smoke test, but I think I have something wrong on my environment and I can't run it (maybe because I'm working on Windows?).
Anyways, I wrote this test:
```
    def test_upgrade_to_chord_inside_chains(self, celery_setup: CeleryTestSetup):
        file_path = tempfile.mktemp()
        try:
            group1 = group(write_to_file_and_return_int.si(file_path, 'a'), write_to_file_and_return_int.si(file_path, 'a'))
            group2 = group(write_to_file_and_return_int.si(file_path, 'a'), write_to_file_and_return_int.si(file_path, 'a'))
            chord1 = group1 | group2
            chain1 = chain(chord1, (write_to_file_and_return_int.si(file_path, 'a') | write_to_file_and_return_int.si(file_path, 'b')))
            chain1.apply_async(queue=celery_setup.worker.worker_queue).get(timeout=RESULT_TIMEOUT)
            with open(file_path, 'r') as f:
                data = f.read()
            assert data.count('b') == 1
        finally:
            pathlib.Path(file_path).unlink(missing_ok=True)
```

**Comment by Nusnus** [comment]
> Hi, I tried to create the smoke test, but I think I have something wrong on my environment and I can't run it (maybe because I'm working on Windows?). Anyways, I wrote this test:
> 
> ```
>     def test_upgrade_to_chord_inside_chains(self, celery_setup: CeleryTestSetup):
>         file_path = tempfile.mktemp()
>         try:
>             group1 = group(write_to_file_and_return_int.si(file_path, 'a'), write_to_file_and_return_int.si(file_path, 'a'))
>             group2 = group(write_to_file_and_return_int.si(file_path, 'a'), write_to_file_and_return_int.si(file_path, 'a'))
>             chord1 = group1 | group2
>             chain1 = chain(chord1, (write_to_file_and_return_int.si(file_path, 'a') | write_to_file_and_return_int.si(file_path, 'b')))
>             chain1.apply_async(queue=celery_setup.worker.worker_queue).get(timeout=RESULT_TIMEOUT)
>             with open(file_path, 'r') as f:
>                 data = f.read()
>             assert data.count('b') == 1
>         finally:
>             pathlib.Path(file_path).unlink(missing_ok=True)
> ```

Thank you for trying! ❤️
Yeah, Windows is not supported, so I guess it was more complicated for you to work on it.

That being said, I’ve tried to help and actually found a bug in the smoke tests env, which I fixed here: https://github.com/celery/celery/pull/9090
After that, and based on the smoke test you tried to implement, I made a smoke test from the integration test of this PR (which is also very similar to your smoke test), which locally seems to pass and we’ll see in the CI very soon if it works too.
```console
tox -e 3.12-smoke -- -k test_upgrade_to_chord_inside_chains -n auto
ROOT: will run in automatically provisioned tox, host /opt/homebrew/Cellar/tox/4.15.1/libexec/bin/python is missing [requires (has)]: tox-gh-actions
ROOT: provision> .tox/.tox/bin/python -m tox -e 3.12-smoke -- -k test_upgrade_to_chord_inside_chains -n auto
ROOT: tox-gh-actions won't override envlist because tox is not running in GitHub Actions
.pkg: _optional_hooks> python /Users/nusnus/dev/GitHub/celery/.tox/.tox/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
.pkg: get_requires_for_build_editable> python /Users/nusnus/dev/GitHub/celery/.tox/.tox/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
.pkg: build_editable> python /Users/nusnus/dev/GitHub/celery/.tox/.tox/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
3.12-smoke: install_package> python -I -m pip install --force-reinstall --no-deps /Users/nusnus/dev/GitHub/celery/.tox/.tmp/package/70/celery-5.4.0-0.editable-py3-none-any.whl
3.12-smoke: commands[0]> pytest -xsvv t/smoke --dist=loadscope --reruns 10 --reruns-delay 60 --rerun-except AssertionError -k test_upgrade_to_chord_inside_chains -n auto
======================================================================================= test session starts =======================================================================================
platform darwin -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0 -- /Users/nusnus/dev/GitHub/celery/.tox/3.12-smoke/bin/python
cachedir: .tox/3.12-smoke/.pytest_cache
rootdir: /Users/nusnus/dev/GitHub/celery
configfile: pyproject.toml
plugins: docker-tools-3.1.3, celery-1.0.0, cov-5.0.0, github-actions-annotate-failures-0.2.0, click-1.1.0, subtests-0.12.1, rerunfailures-14.0, order-1.2.1, anyio-4.4.0, timeout-2.3.1, xdist-3.6.1
10 workers [2 items]
scheduling tests via LoadScopeScheduling

t/smoke/tests/test_canvas.py::test_chain::test_upgrade_to_chord_inside_chains[celery_setup_worker-celery_redis_broker-celery_redis_backend]
[gw0] PASSED t/smoke/tests/test_canvas.py::test_chain::test_upgrade_to_chord_inside_chains[celery_setup_worker-celery_redis_broker-celery_redis_backend]
t/smoke/tests/test_canvas.py::test_chain::test_upgrade_to_chord_inside_chains[celery_setup_worker-celery_rabbitmq_broker-celery_redis_backend]
[gw0] PASSED t/smoke/tests/test_canvas.py::test_chain::test_upgrade_to_chord_inside_chains[celery_setup_worker-celery_rabbitmq_broker-celery_redis_backend]

======================================================================================= 2 passed in 41.53s ========================================================================================
  3.12-smoke: OK (44.50=setup[1.64]+cmd[42.86] seconds)
  congratulations :) (46.77 seconds)
```

If everything passes successfully, I’ll merge the PR.
It should have a passing unit, integration and now smoke tests to confirm the bug is fully fixed.

Now let’s wait for the CI to be over - got my :eye

**Pull request merged**
